### PR TITLE
Get the value from SearchInput and pass it to the fetch url in App.js

### LINKED ISSUE #1: Value from search input field should be added to the Fetch url i.e ?term=[value goes here]
I need to make sure that the <SearchInput /> component takes the value and adds it to the url in the url of the fetch() which currently sits in the App.js file. 

- [x] Get the value of the input field into `<SearchInput />`'s state. 
- [ ] Get the value from state of `<SearchInput />` into the `<App />` component. 
- [ ] Check that the search term renders the correct jokes to the canvas.

### PR DESCRIPTION
**Pull request by markwillis**
Fixes #1 

- [x] Get the search value into state
- [x] Check it works by console logging `this.state.searchTerm` on formSubmit
- [x] Pass this value to the App.js file
- [x] Insert it into the fetch url

I'm struggling with the last two...

### CODE REVIEW

File: src/components/SearchInput.jsx [hunk 1]
```diff
@@ -8,22 +8,23 @@ class SearchInput extends React.Component {
...
   }
 
   handleChange(event) {
     this.setState({searchTerm: event.target.value})
-    console.log(this.state.searchTerm)
   }
 
   handleClick(event) {
     event.preventDefault()
+    console.log(this.state.searchTerm)
```
Comment (markwillis) [inline src/components/SearchInput.jsx]: When I use the `Amuse me` button, I'm able to log state to the console. 

I'm really struggling on how to pass it either as props, or as a variable to import into App.js.

**Comment by markwillis** [comment]
The search input is working! Woohoo!

Ready for a code review on this @arb 😬

File: src/App.js [hunk 2]
```diff
@@ -1,43 +1,87 @@
...
-    this.state = {
-      jokeData: {
-        results: []
-      }
+  state = {
+    jokeData: {
+      results: []
+    },
+    searchTerm: "",
+    loadingState: "resting"
```
Comment (arb) [inline src/App.js]: It would be a good idea to turn the possible values of `loadingState` into an enumeration so you don't need to make sure you type and spell it correctly everywhere you use it.

File: src/App.js [hunk 3]
```diff
@@ -1,43 +1,87 @@
...
+  };
+
+  handleChange = event => {
+    this.setState({ searchTerm: event.target.value });
```
Comment (arb) [inline src/App.js]: You should use `currentTarget` here rather than `target`. And before you ask, [this](https://stackoverflow.com/questions/10086427/what-is-the-exact-difference-between-currenttarget-property-and-target-property) gives a good answer to the difference as well as a link out to a blog post explaining it.

File: src/App.js [hunk 4]
```diff
@@ -1,43 +1,87 @@
...
+  };
+
+  handleClick = event => {
+    event.preventDefault();
+    this.setState({ loadingState: "loading" });
```
Comment (arb) [inline src/App.js]: So if you create that enum I was talking about earlier, this would be `loadingState: STATE.loading` instead.

File: src/App.js [hunk 5]
```diff
@@ -1,43 +1,87 @@
...
+  };
+
+  async fetchJokes() {
+    try {
+      const url =
```
Comment (arb) [inline src/App.js]: I'd probably exit early if the search term is empty. Also, you may want to add a check for annoying crap like a single space or like 20 spaces.

File: src/App.js [hunk 6]
```diff
@@ -1,43 +1,87 @@
...
+        this.state.searchTerm === ""
+          ? `https://icanhazdadjoke.com/search`
+          : `https://icanhazdadjoke.com/search?term=${this.state.searchTerm}`;
```
Comment (arb) [inline src/App.js]: This is probably ok for a first version, but what happens if they search for `"cats and dogs"`? What you have here won't set `term` correctly.

You may want to look into something like https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams for more robust query string creation.

File: src/App.js [hunk 7]
```diff
@@ -1,43 +1,87 @@
...
+      const fetchUrl = fetch(url, {
```
Comment (arb) [inline src/App.js:46]: This is fine as-is, but you could shorten it to something like
```
const data = await (fetch(url, {
  method: "GET",
  headers: { accept: "application/json" }
})).json();
```

I didn't test that, but that or something VERY similar would save you a few lines.

File: src/App.js [hunk 8]
```diff
@@ -1,43 +1,87 @@
...
+        method: "GET",
+        headers: { accept: "application/json" }
```
Comment (arb) [inline src/App.js]: Since the value of `headers` never changes, you could pull it out of `fetchJokes` and just declare it once outside of the component. Every time `fetchJokes` is called, it creates a new object. Since it's value never changes, it's safe to keep reusing the same one. Every time you create a new object, it uses memory that will need to be garbage collected eventually. 

For something this small, it really doesn't matter, but I'm trying to help you learn beyond this code 😅

File: src/App.js [hunk 9]
```diff
@@ -1,43 +1,87 @@
...
+      });
+      const response = await fetchUrl;
+      const myData = await response.json();
+      this.setState({
+        jokeData: { results: myData.results },
+        loadingState: "loaded"
+      });
+    } catch (e) {
+      this.setState({ loadingState: "Something went wrong" });
```
Comment (arb) [inline src/App.js]: You're overloading `loadingState` here. Everywhere else, it's a general state of the component but here it's a message you're likely displaying to the user. I think you should make the loading state `STATE.error` or something and then display a message in `render` based on `this.state.loadingState`.

File: src/App.js [hunk 10]
```diff
@@ -1,43 +1,87 @@
...
-  async componentDidMount() {
-    const url = 'https://icanhazdadjoke.com/search?term="cat"';
-    const fetchUrl = fetch(url, {
-      method: "GET",
-      headers: { accept: "application/json" }
-    });
-    const response = await fetchUrl;
-    const myData = await response.json();
-    this.setState({ jokeData: { results: myData.results } });
+  componentDidUpdate(prevProps, prevState) {
```
Comment (arb) [inline src/App.js]: Any time I see `componentDidUpdate` being used it makes me double check... you sure you need this? What problem are you trying to solve by adding this?

File: src/App.js [hunk 11]
```diff
@@ -1,43 +1,87 @@
...
+    ));
     return (
       <div className="App">
-        <PageHeader />
+        <PageHeader>
+          <SearchInput
+            handleChange={this.handleChange}
+            handleClick={this.handleClick}
+            placeholder="Search for a joke..."
+            onChange={this.handleChange}
```
Comment (arb) [inline src/App.js]: I knew you'd figure this out eventually! Nice job 💯

File: src/components/SearchInput.jsx [hunk 12]
```diff
@@ -1,45 +1,20 @@
...
-    );
-  }
-}
-
-SearchInput.propTypes = {
-  searchTerm: PropTypes.string
+function SearchInput(props) {
+  
+  return (
+    <form onSubmit={props.handleClick}>
```
Comment (arb) [inline src/components/SearchInput.jsx]: These prop names are a little confusing. If I was a consumer passing in a `handleClick` prop, I wouldn't expect it to be used when a form is submitted.

File: src/App.js [hunk 13]
```diff
@@ -1,43 +1,87 @@
...
+  }
+
+  componentDidMount() {
+    this.fetchJokes();
   }
 
   // This needs rendering inside a map function, then rendering in the render as <div>{jokeList}</div>
   // <Joke key="[ENTER THE KEY HERE]" joke={this.state}/>
   render() {
     const jokeList = this.state.jokeData.results.map(item => (
```
Comment (arb) [inline src/App.js]: I know why you did it this way, but this causes this variable to created and for the iterating inside `map` to happen even when `loadingState` isn't "loaded". Usually these kinds of things are "in-lined" down on line 75.

File: src/App.js [hunk 14]
```diff
@@ -1,43 +1,87 @@
...
+          {this.state.loadingState === "loaded"
+            ? jokeList
+            : `${this.state.loadingState}`}
         </div>
       </div>
     );
   }
 }
 
+SearchInput.propTypes = {
```
Comment (arb) [inline src/App.js]: If you're going to use `propTypes` you should set them up where `SearchInput` is declared ("SearchInput.jsx " in this case). Doing it this way is confusing and potentially dangerous. You could be overriding the existing `propTypes` set up in "SearchInput.jsx "

File: src/components/PageHeader.jsx [hunk 15]
```diff
@@ -1,13 +1,13 @@
...
 
 function PageHeader(props) {
+
   return (
     <div className="hero">
       <h1>
         The <span>ULTIMATE</span> Dad joke collection
       </h1>
-      <SearchInput />
+      {props.children}
```
Comment (arb) [inline src/components/PageHeader.jsx:10]: Smart 👍 Sometimes takes people a LONG time to figure that out.

**Review by arb — CHANGES_REQUESTED** [review]
This looks great so far! A lot of just small ideas and suggestions but you got it working and that's awesome buddy!!

File: src/components/SearchInput.jsx (hunk 12 above)
Comment (markwillis) [reply src/components/SearchInput.jsx]: Ah, handleSubmit...

File: src/App.js (hunk 2 above)
Comment (markwillis) [reply src/App.js]: So would I add a PropType here such as
`loadingState: PropTypes.oneOf(['resting', 'loaded', 'loading']).isRequired`

and then set a variable such as
`const message = {resting: "resting", loaded: "loaded"}` etc

So that way I could use `state = {loadingState: {loaded: message.loaded, resting: message.resting}` etc?

File: src/App.js (hunk 4 above)
Comment (markwillis) [reply src/App.js]: Ah. Answers my first question 👍

File: src/App.js (hunk 3 above)
Comment (markwillis) [reply src/App.js]: Thanks for adding a link here. Makes perfect sense in this case.

File: src/App.js (hunk 2 above)
Comment (arb) [reply src/App.js]: Kind of... since App.js is the main entry point into your application, nothing really passes any props into it, so don't worry about PropTypes.

An enum (enumeration) would look something like this:
```
const LOADING_STATE = { RESTING: 'resting', LOADED: 'loaded', LOADING: 'loading', ...}
```
How you case the values is up to you and really doesn't matter, what matters is that you don't use "magic strings" throughout the application.

File: src/App.js (hunk 5 above)
Comment (markwillis) [reply src/App.js]:
```js
const regex = /^[a-zA-Z0-9_]+$/
    try {
      const url =
        regex.test(this.state.searchTerm)
          ? `https://icanhazdadjoke.com/search?term=${this.state.searchTerm}`
          : `https://icanhazdadjoke.com/search`;
      const fetchUrl = fetch(url, {
```

So I added this, which works and stops any empty strings being added...
I wanted to add style to the input to show there is an error, perhaps adding a value to state like `{valid: true}` which is updated if the regex is valid or not. 

Would that be the right path here?

File: src/App.js (hunk 14 above)
Comment (markwillis) [reply src/App.js]: They're gone for now

File: src/App.js (hunk 10 above)
Comment (markwillis) [reply src/App.js]: In the handleSubmit, the state is set to `loading` when the user submits the search. In the `componentDidUpdate()` if Loading state = loading, then it fetches those jokes agian with necessary term appended.

File: src/App.js (hunk 7 above)
Comment (markwillis) [reply src/App.js:46]: I'm not sure where to put the repsonse here and can't get it to work.

I tried
``` js
async fetchJokes() {
    this.checkInput(this.state.searchTerm)
    try {
      const myData = await (fetch(url, {
        method: "GET", headers: headers
      })).json();
      this.setState({
        jokeData: { results: myData.results },
        loadingState: LOADING_STATE.loaded
      });
```
So I'll park it for now and come back to it.

File: src/components/Footer.jsx [hunk 16]
```diff
@@ -0,0 +1,14 @@
+import React from 'react'
+
+function Footer() {
+  return (
+    <footer className="footer">
+      <div>
+        <h3>Thanks for visiting.</h3>
+        <p>Built with React.js using icanhazdadjoke API © {new Date().getFullYear()}</p>
```
Comment (arb) [inline src/components/Footer.jsx:8]: So this is kind of an advanced technique but... the problem with `Footer` getting the full year itself is that it makes testing it a hassle. One of the programming models around react is "given the exact same props, you'll get the same rendered output". The way `Footer` is written, that is no longer true. It renders differently in 2019 compared to 2020 which very recently just happened. This is called a "side effect". This code "does something" that impacts the output that the consumer (App.jsx) can't see or control.

TL;DR, pass in `now` or `year` as a prop. Then you'll have control from OUTSIDE the component if/when you end up writing a test.

File: src/App.js [hunk 17]
```diff
@@ -1,43 +1,82 @@
...
-    return (
-      <div className="App">
-        <PageHeader />
-        <div className="joke-wrapper">
-          {jokeList}
-        </div>
+  useEffect(() => {
+    fetchJokes()
+  }, [])
+
```
Comment (markwillis) [inline src/App.js:61]: I"m getting Lint errors here that I don't really undertand. It's suggesting I use `useCallback` function so that it will update if either one of the array items is updated. I kind of get why, but I don't understand the syntax to use here.

File: src/App.js [hunk 18]
```diff
@@ -1,43 +1,82 @@
...
+  const [jokeData, setJokeData] = useState({ })
+  const [searchTerm, setSearchTerm] = useState("hat")
+  const [loadingState, setLoadingState] = useState(LOADING_STATE.resting)
+  
+
+  function handleChange(event) {
+    const {value} = event.target
+    setSearchTerm(value);
+    console.log("hello")
+  };
```
Comment (markwillis) [inline src/App.js:27]: @arb This handlechange is no longer working. Well it's not console logging anything when typing in the field and I'm not sure why....
